### PR TITLE
#3295 - Add log redaction infrastructure

### LINKED ISSUE #3295: Infrastructure to support redacted logging
- [ ] Add logging w/ redaction API to logging.go
- [ ] Define redacted types 
- [ ] Unit testing

### PR DESCRIPTION
**Pull request by bbrks**
Closes #3295 

- [x] Add Redactor interface
- [x] Add UserData type implementing Redactor
- [x] Add redacted log functions
- [x] Unit testing

Optimised for zero-allocs:

```
$ go test -run='(Redact|UD)' -bench='Redact' -benchmem -v ./base
=== RUN   TestRedactedLogFuncs
--- PASS: TestRedactedLogFuncs (0.00s)
=== RUN   TestRedactHelper
--- PASS: TestRedactHelper (0.00s)
=== RUN   TestUserDataRedact
--- PASS: TestUserDataRedact (0.00s)
=== RUN   TestUD
--- PASS: TestUD (0.00s)
goos: darwin
goarch: amd64
pkg: github.com/couchbase/sync_gateway/base
BenchmarkRedactHelper-8     	20000000	        59.5 ns/op	       0 B/op	       0 allocs/op
BenchmarkUserDataRedact/Enabled-8         	50000000	        28.8 ns/op	       0 B/op	       0 allocs/op
BenchmarkUserDataRedact/Disabled-8        	2000000000	         0.54 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/couchbase/sync_gateway/base	3.881s
```

### CODE REVIEW

File: base/log_redactor_userdata.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+)
+
+// RedactUserData is a global toggle for UserData redaction.
+var RedactUserData = false
+
+// UserData implements the Redactor interface for logging purposes.
+type UserData string
+
+// Compile-time interface check.
+var _ Redactor = UserData("")
```
Comment (bbrks) [inline base/log_redactor_userdata.go]: Here I'm forcing the compiler to check `UserData` satisfies `Redactor`. This is more useful in scenarios where the interface and implementation are in separate packages, but still a useful thing to have around IMO.

https://medium.com/@matryer/golang-tip-compile-time-checks-to-ensure-your-type-satisfies-an-interface-c167afed3aae

File: base/redactor.go [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+type Redactor interface {
+	// Redact returns the given string in a redacted form. This may be tagged,
+	// changed, hashed, or removed completely depending on desired behaviour.
+	Redact() string
+}
+
+// Redact performs an *in-place* redaction on the input slice, and returns it.
+// This should only be consumed by logging funcs. E.g. fmt.Printf(fmt, Redact(args))
+//
+// FIXME: Ideally this would be private, but we'd need to pull 'base/logging.go' into this package.
```
Comment (adamcfraser) [inline base/redactor.go]: Comment update

File: base/logging.go [hunk 3]
```diff
@@ -236,6 +237,19 @@ func SetLogLevel(level int) {
...
+	case RedactFull:
+		RedactUserData = true
+	case RedactPartial:
+		RedactUserData = true
+	case RedactNone:
+	default:
+		RedactUserData = false
+	}
+}
+
```
Comment (adamcfraser) [inline base/logging.go]: Add a unit test that loads from config and validates RedactUserData ends up set properly.

**Review by adamcfraser — COMMENTED** [review]
Looks good - couple of minor comments.

**Review by adamcfraser — APPROVED** [review]
Looks good

**Pull request merged**